src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts

```diff
@@ -544,15 +604,189 @@ describe('AdminSearchPageComponent', () => {
     regenerateButton.click();
 
     expect(spyStatusMessageService).toBeCalled();
   });
 
+  it('should generate email when course join email button clicked', () => {
+    const studentResult: StudentAccountSearchResult = {
+      name: 'jack',
+      email: 'email@test.com',
+      googleId: 'googleId',
+      courseId: 'courseId',
+      courseName: 'courseName',
+      institute: 'institute',
+      courseJoinLink: 'courseJoinLink',
+      homePageLink: 'homePageLink',
+      manageAccountLink: 'manageAccountLink',
+      showLinks: true,
+      section: 'section',
+      team: 'team',
+      comments: 'comments',
+      recordsPageLink: 'recordsPageLink',
+      openSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+      notOpenSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+      publishedSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+    };
+    component.students = [studentResult];
+    fixture.detectChanges();
+
+    const spyEmailGenerationService: any = spyOn(emailGenerationService, 'getCourseJoinEmail')
+        .and.callFake(
+            (): Observable<Email> => of({
+              recipient: 'Jacky Chan',
+              subject: 'Course join email',
+              content: 'Course join email content',
+            }),
+            );
+
+    const sendButton: any = fixture.debugElement.nativeElement.querySelector('#send-course-join-button');
+    sendButton.click();
+
+    expect(spyEmailGenerationService).toBeCalled();
+  });
+
+  it('should show error message if fail to send course join email', () => {
+    const studentResult: StudentAccountSearchResult = {
+      name: 'name',
+      email: 'email',
+      googleId: 'googleId',
+      courseId: 'courseId',
+      courseName: 'courseName',
+      institute: 'institute',
+      courseJoinLink: 'courseJoinLink',
+      homePageLink: 'homePageLink',
+      manageAccountLink: 'manageAccountLink',
+      showLinks: true,
+      section: 'section',
+      team: 'team',
+      comments: 'comments',
+      recordsPageLink: 'recordsPageLink',
+      openSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+      notOpenSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+      publishedSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+    };
+    component.students = [studentResult];
+    fixture.detectChanges();
+
+    spyOn(emailGenerationService, 'getCourseJoinEmail').and.returnValue(throwError({
+      error: {
+        message: 'This is the error message.',
+      },
+    }));
+
+    const spyStatusMessageService: any = spyOn(statusMessageService, 'showErrorToast').and.callFake((args: string) => {
+      expect(args).toEqual('This is the error message.');
+    });
+
+    const sendButton: any = fixture.debugElement.nativeElement.querySelector('#send-course-join-button');
+    sendButton.click();
+
+    expect(spyStatusMessageService).toBeCalled();
+  });
+
+  it('should generate email when send session reminder email button clicked', () => {
+    const studentResult: StudentAccountSearchResult = {
+      name: 'Jack Chan',
```

Comment:
Could we abstract the `StudentAccountSearchResult` objects in this file into a factory function of sorts, and you can populate the necessary fields as parameters. Seems to me like not many fields need to change between test cases. For example, 

```
const initStudentSearchResult = (name, email) => {
  return {
    // populate data

Reply:
@madanalogy Thanks for the suggestion, made the changes 👍